### PR TITLE
Add Reflection Guide

### PR DESCRIPTION
**Pull request by gnossen**

### CODE REVIEW

File: content/en/docs/guides/reflection.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+One coming from the REST world might compare the gRPC reflection API to serving
+an OpenAPI document on the HTTP server presenting the REST API being described.
+
+### Transparency and Interpretability
+
+A big contributor to gRPC's stellar performance is the use of Protobuf for
+serialization -- a _binary_ non-human-readable protocol. While this greatly
+speeds up an RPC, it can also make it more difficult to manually interact with a
+server. In order to manually send a gRPC request to a server over HTTP/2 using
+`curl`, you would have to:
```
Comment (temawi) [inline content/en/docs/guides/reflection.md:29]: Maybe add "Hypothetically, " at the beginning of this sentence to make it clear this si not actually something anyone should attempt.

File: content/en/docs/guides/reflection.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+[Python example]: https://github.com/grpc/grpc/blob/master/examples/python/helloworld/greeter_server_with_reflection.py
+
+### Tips
+ 
+Reflection works so seamlessly with tools such as `grpcurl` that oftentimes,
+people aren't even aware that it's happening under the hood. However, if
+reflection isn't exposed, things won't work seamlessly at all. Instead, the
+client will fail with nasty errors. People often run into this when writing
+routing configuration for a gRPC service. The _reflection_ service must be
```
Comment (temawi) [inline content/en/docs/guides/reflection.md:72]: "writing a routing configuration"

**Review by temawi — COMMENTED** [review]
Looks great!

File: content/en/docs/guides/reflection.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+description: >-
+  Explains how reflection can be used to improve the transparency and
+  interpretability of RPCs.
+---
+
+### Overview
+
+Reflection is [a
+protocol](https://github.com/grpc/grpc-proto/blob/master/grpc/reflection/v1/reflection.proto)
+that gRPC servers can use to declare the APIs they export over a standardized
```
Comment (ejona86) [inline content/en/docs/guides/reflection.md:12]: s/the APIs/the protobuf APIs/

File: content/en/docs/guides/reflection.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+1. Know which RPC services the server exposed.
+2. Know the protobuf definition of the request message and all types it
+references.
+3. Know the protobuf definition of the response message all the types _it_
+references.
+
+Then, you'd have to use that knowledge to hand-craft your request message(s) into
+binary and painstakingly decode the response message(s). This would be time
+consuming, frustrating, and error prone. Instead, the reflection protocol
```
Comment (ejona86) [inline content/en/docs/guides/reflection.md:39]: "Instead, the reflection protocol allows tools to..."

**Review by ejona86 — COMMENTED** [review]
Just two nits. I'll let Terry be the approver.

File: content/en/docs/guides/reflection.md (hunk 1 above)
Comment (gnossen) [reply content/en/docs/guides/reflection.md]: Done.

File: content/en/docs/guides/reflection.md (hunk 2 above)
Comment (gnossen) [reply content/en/docs/guides/reflection.md:72]: Done.

File: content/en/docs/guides/reflection.md (hunk 3 above)
Comment (gnossen) [reply content/en/docs/guides/reflection.md]: Done

File: content/en/docs/guides/reflection.md (hunk 4 above)
Comment (gnossen) [reply content/en/docs/guides/reflection.md:39]: Done.

**Comment by gnossen** [comment]
@temawi PTAL.

**Review by temawi — APPROVED** [review]

**Pull request merged**